synapse/storage/devices.py

```diff
@@ -97,20 +97,69 @@ def get_devices_by_remote(self, destination, from_stream_id):
     def _get_devices_by_remote_txn(
         self, txn, destination, from_stream_id, now_stream_id
     ):
+        # We retrieve n+1 devices from the list of outbound pokes were n is our
+        # maximum. We then check if the very last device has the same stream_id as the
+        # second-to-last device. If so, then we ignore all devices with that stream_id
+        # and only send the devices with a lower stream_id.
+        #
+        # If when culling the list we end up with no devices afterwards, we consider the
+        # device update to be too large, and simply skip the stream_id - the rationale
+        # being that such a large device list update is likely an error.
+        #
+        # Note: The code below assumes this value is at least 1
+        maximum_devices = 100
```

Comment:
this should probably be passed in as a `limit` param, so that it can be derived from https://github.com/matrix-org/synapse/blob/develop/synapse/federation/sender/per_destination_queue.py#L37.